source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,487 @@
+        }
+        return entry;
+      }
+
+      return findOrCreateSubset(entry->children_, kvs, idx, allow_create);
+    }
+  }
+
+  // not found, but don't want creation
+  if (!allow_create) {
+    return nullptr;
+  }
+
+  LbSubsetEntryPtr entry(new LbSubsetEntry());
+  if (kv_it != subsets.end()) {
+    ValueSubsetMap& value_subset_map = kv_it->second;
+    value_subset_map.emplace(value, entry);
+  } else {
+    ValueSubsetMap value_subset_map = {{value, entry}};
+    subsets.emplace(name, value_subset_map);
+  }
+
+  idx++;
+  if (idx == kvs.size()) {
+    entry->host_subset_ =
+        newSubset(std::bind(&SubsetLoadBalancer::hostMatches, this, kvs, std::placeholders::_1));
+    return entry;
+  }
+
+  return findOrCreateSubset(entry->children_, kvs, idx, true);
+}
+
+// Create a new Subset from the current HostSets, filtering hosts with
+// the given predicate.
+SubsetLoadBalancer::SubsetPtr
+SubsetLoadBalancer::newSubset(std::function<bool(const HostSharedPtr&)> predicate) {
+  std::vector<HostSharedPtr> hosts;
+  for (const auto& host : original_host_set_.hosts()) {
+    if (predicate(host)) {
+      hosts.emplace_back(host);
+    }
+  }
+
+  SubsetPtr subset;
+  std::vector<HostSharedPtr> local_hosts;
+  if (original_local_host_set_) {
+    for (const auto& host : original_local_host_set_->hosts()) {
+      if (predicate(host)) {
+        local_hosts.emplace_back(host);
+      }
+    }
+
+    subset.reset(new Subset(original_host_set_, *original_local_host_set_));
+  } else {
+    subset.reset(new Subset(original_host_set_, empty_host_set_));
+  }
+
+  subset->host_set_.update(hosts, {});
+  subset->local_host_set_.update(local_hosts, {});
+  return subset;
+}
+
+LoadBalancer* SubsetLoadBalancer::newLoadBalancer(const SubsetPtr& subset) {
+  const HostSet* local_host_set = nullptr;
+  if (original_local_host_set_ != nullptr) {
+    local_host_set = &subset->local_host_set_;
+  }
+
+  LoadBalancer* lb = nullptr;
+  switch (lb_type_) {
+  case LoadBalancerType::LeastRequest:
+    lb = new LeastRequestLoadBalancer(subset->host_set_, local_host_set, stats_, runtime_, random_);
+    break;
+
+  case LoadBalancerType::Random:
+    lb = new RandomLoadBalancer(subset->host_set_, local_host_set, stats_, runtime_, random_);
+    break;
+
+  case LoadBalancerType::RoundRobin:
+    lb = new RoundRobinLoadBalancer(subset->host_set_, local_host_set, stats_, runtime_, random_);
+    break;
+
+  case LoadBalancerType::RingHash:
+    lb = new RingHashLoadBalancer(subset->host_set_, stats_, runtime_, random_);
+    break;
+
+  case LoadBalancerType::OriginalDst:
+    NOT_REACHED;
+  }
+
+  subset->host_set_.triggerCallbacks();
+  if (local_host_set) {
+    subset->local_host_set_.triggerCallbacks();
+  }
+  return lb;
+}
+
+// Find the index in which the original HostSet's hostsPerLocality vector
+// contains the given host.
+int SubsetLoadBalancer::HostSubsetImpl::findLocalityIndex(const HostSharedPtr& host) {
```

Comment:
Independent of this PR, I think the `HostSet` locality relationship could be improved  to provide more direct mapping between a host and its locality equivalence class. Right now, it's just this vector of `HostSets`, but it could be a map from `Locality` to `HostSet`, and each `Host` knows its own `Locality` to index with.